dice-engine/src/dice_engine/dice.clj

```diff
@@ -1,40 +1,127 @@
-(ns dice-engine.dice)
+(ns dice-engine.dice
+  (:refer-clojure :exclude [drop keep]))
+
+(def id-counter (atom 0))
+
+(defn get-id
+  "Returns an auto incremented id"
+  []
+  (swap! id-counter inc))
+
+(defn rand-int-natural
+  "Returns a random integer between 1 to n (both inclusive)"
+  [n]
+  (+ 1 (rand-int n)))
+
+(defn create-die
+  "Returns a die map of the form
+  {:id              (get-id)
+   :value           n
+   :faces           x
+   :previous-values []}"
+  [die-value faces]
+  {:id              (get-id)
+   :value           die-value
+   :faces           faces
+   :previous-values []})
+
+(defn sum
+  "Takes a list of die maps and returns the sum of the values"
+  [dice]
+  (->>
+    dice
+    (map :value)
+    (reduce +)))
+
+(defn roll
+  "Returns n dice, each with a value <= faces"
+  [n faces]
+  (->>
+    (repeatedly n #(rand-int-natural faces))
+    (map #(create-die % faces))))
+
+(defn reroll
+  "Rerolls a die"
+  [{:keys [value faces previous-values] :as die}]
+  (assoc die :value (rand-int-natural faces)
+             :previous-values (cons value previous-values)))
 
 (defn drop
-  "Takes a dice map, discards dice that match n and returns a new dice map."
-  [dice-roll & n]
-  ())
+  "Applies the selector on dice with literal.
+  Returns dice unselected by selector."
+  [selector literal dice]
+  (->>
+    dice
+    (selector literal)
+    (second)))
 
 (defn keep
-  "Takes a dice map, keeps dice that match n and returns a new dice map."
-  [dice-roll & n]
-  )
+  "Applies the selector on dice with literal.
+  Returns dice selected by selector."
+  [selector literal dice]
+  (->>
+    dice
+    (selector literal)
+    (first)))
 
-(defn reroll
-  "Takes a dice map, re-rolls dice that match n.
-  Returns a new dice map if none of the rerolled dice match n,re-rolls until none match.
-  Updates history of each die on every roll."
-  [dice-roll & n]
-  )
+(defn reroll-matched
+  "Rerolls dice selected by selector and returns the rerolled dice.
+   Rerolls until none of the dice qualify the selector."
+  [selector literal dice]
+  (let [[matched-dice rest-dice] (selector literal dice)]
+    (if (empty? matched-dice)
+      dice
+      (recur
+        selector
+        literal
+        (concat (map reroll matched-dice) rest-dice)))))
```

Comment:
`reroll-matched` and other filtering functions (`keep`, `drop`) don't need to return the result values in order.

The calling function `evaluate-roll` keeps track of the initial order. And since each die has an id, they can later be output in the order they were rolled.